server/src/main/java/org/elasticsearch/action/index/IndexRequest.java

```diff
@@ -103,6 +106,15 @@ public class IndexRequest extends ReplicatedWriteRequest<IndexRequest> implement
     private boolean isPipelineResolved;
 
     private boolean requireAlias;
+    /**
+     * This indicates whether the response to this request ought to list the ingest pipelines that were executed on the document
+     */
+    private boolean listExecutedPipelines;
+    /**
+     * This holds the names of the ingest pipelines that have been executed on the document for this request. This is not meant to be set by
+     * the creator of the request -- pipelines are added here at runtime as they are executed.
+     */
+    private List<String> executedPipelines = new ArrayList<>();
```

Comment:
I'd like this to be closer to cost-free in the case of it not being used.

Reply:
Consider the very similar (but not identical) `private final Set<String> executedPipelines = new LinkedHashSet<>();` that we're already tracking on IngestDocument itself. If these end up always having the same details, perhaps we could have this new feature just serialize off the thing we're already tracking rather than tracking a new thing?

Reply:
> I'd like this to be closer to cost-free in the case of it not being used.

Good point. I've updated it to only allocate that memory when the first pipeline is added if `listExecutedPipelines` is set to true (false by default). I also changed the serialization code so that if `listExecutedPipelines` is false then we don't serialize the collection at all, so now the serialization cost is only one byte per doc.

> Consider the very similar (but not identical) private final Set<String> executedPipelines = new LinkedHashSet<>(); that we're already tracking on IngestDocument itself

I don't think that that's reusable here, because it's tracking which pipelines are currently in use. Completed pipelines are removed.